src/algebra/ordered_group.lean

```diff
@@ -696,43 +865,71 @@ instance linear_ordered_comm_group.to_no_bot_order [nontrivial α] : no_bot_orde
 
 end linear_ordered_comm_group
 
-section linear_ordered_add_comm_group
+section covariant_add_le
 
-variables [linear_ordered_add_comm_group α] {a b c : α}
+section has_neg
+variables [has_neg α] [linear_order α] {a b: α}
 
-@[simp]
-lemma sub_le_sub_flip : a - b ≤ b - a ↔ a ≤ b :=
+/-- `abs a` is the absolute value of `a`. -/
+def abs {α : Type*} [has_neg α] [linear_order α] (a : α) : α := max a (-a)
+
+lemma abs_choice (x : α) : abs x = x ∨ abs x = -x := max_choice _ _
+
+lemma abs_le' : abs a ≤ b ↔ a ≤ b ∧ -a ≤ b := max_le_iff
+
+lemma le_abs : a ≤ abs b ↔ a ≤ b ∨ a ≤ -b := le_max_iff
+
+lemma le_abs_self (a : α) : a ≤ abs a := le_max_left _ _
+
+lemma neg_le_abs_self (a : α) : -a ≤ abs a := le_max_right _ _
+
+lemma lt_abs : a < abs b ↔ a < b ∨ a < -b := lt_max_iff
+
+theorem abs_le_abs (h₀ : a ≤ b) (h₁ : -a ≤ b) : abs a ≤ abs b :=
+(abs_le'.2 ⟨h₀, h₁⟩).trans (le_abs_self b)
+
+lemma abs_by_cases (P : α → Prop) {a : α} (h1 : P a) (h2 : P (-a)) : P (abs a) :=
+sup_ind _ _ h1 h2
+
+end has_neg
+
+section add_group
+variables [add_group α] [linear_order α]
+
+@[simp] lemma abs_neg (a : α) : abs (-a) = abs a :=
+begin unfold abs, rw [max_comm, neg_neg] end
+
+lemma abs_sub (a b : α) : abs (a - b) = abs (b - a) :=
+by rw [← neg_sub, abs_neg]
+
+lemma eq_or_eq_neg_of_abs_eq {a b : α} (h : abs a = b) : a = b ∨ a = -b :=
+by simpa only [← h, eq_comm, eq_neg_iff_eq_neg] using abs_choice a
+
+lemma abs_eq_abs {a b : α} : abs a = abs b ↔ a = b ∨ a = -b :=
 begin
-  rw [sub_le_iff_le_add, sub_add_eq_add_sub, le_sub_iff_add_le],
-  split,
-  { intro h,
-    by_contra H,
-    rw not_le at H,
-    apply not_lt.2 h,
-    exact add_lt_add H H, },
-  { intro h,
-    exact add_le_add h h, }
+  refine ⟨λ h, _, λ h, _⟩,
+  { obtain rfl | rfl := eq_or_eq_neg_of_abs_eq h;
+    simpa only [neg_eq_iff_neg_eq, neg_inj, or.comm, @eq_comm _ (-b)] using abs_choice b },
+  { cases h; simp only [h, abs_neg] },
 end
 
+variables [covariant_class α α (+) (≤)] {a b c : α}
+
 lemma le_of_forall_pos_le_add [densely_ordered α] (h : ∀ ε : α, 0 < ε → a ≤ b + ε) : a ≤ b :=
 le_of_forall_le_of_dense $ λ c hc,
-calc a ≤ b + (c - b) : h _ (sub_pos_of_lt hc)
-   ... = c           : add_sub_cancel'_right _ _
+calc a ≤ b + (-b + c) : h _ (lt_neg_iff_lt.mpr hc)
+   ... = c            : add_neg_cancel_left b c
+
+lemma le_of_forall_pos_lt_add (h : ∀ ε : α, 0 < ε → a < b + ε) : a ≤ b :=
+le_of_not_lt (λ h₁, lt_irrefl a (by simpa using (h _ (lt_neg_iff_lt.mpr h₁))))
 
 lemma le_iff_forall_pos_le_add [densely_ordered α] : a ≤ b ↔ ∀ ε, 0 < ε → a ≤ b + ε :=
 ⟨λ h ε ε_pos, le_add_of_le_of_nonneg h ε_pos.le, le_of_forall_pos_le_add⟩
 
-lemma le_of_forall_pos_lt_add (h : ∀ ε : α, 0 < ε → a < b + ε) : a ≤ b :=
-le_of_not_lt $ λ h₁, by simpa using h _ (sub_pos_of_lt h₁)
 
 lemma le_iff_forall_pos_lt_add : a ≤ b ↔ ∀ ε, 0 < ε → a < b + ε :=
 ⟨λ h ε, lt_add_of_le_of_pos h, le_of_forall_pos_lt_add⟩
```

Comment:
can you move these lemmas not about `abs` to before the definition of `abs`?

Reply:
I moved these lemmas and I also proved the multiplicative version and got `to_additive` to create the additive one.